presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java

```diff
@@ -0,0 +1,543 @@
+         */
+        private synchronized BufferResult processRead(long sequenceId, DataSize maxSize)
+        {
+            // if request is for pages before the current position, just return an empty result
+            // if it is for pages after the current position, there was an error somewhere, so return
+            // and empty result also, and the error will be resolved when the client reconnects
+            if (sequenceId != currentSequenceId.get()) {
+                return emptyResults(taskInstanceId, sequenceId, false);
+            }
+
+            // if buffer is finished, return an empty result
+            if (finished.get()) {
+                return emptyResults(taskInstanceId, currentSequenceId.get(), true);
+            }
+
+            // read the new pages
+            long maxBytes = maxSize.toBytes();
+            List<Page> result = new ArrayList<>();
+            long bytes = 0;
+
+            for (Page page : pages) {
+                bytes += page.getSizeInBytes();
+                // break (and don't add) if this page would exceed the limit
+                if (!result.isEmpty() && bytes > maxBytes) {
+                    break;
+                }
+                result.add(page);
+            }
+            return new BufferResult(taskInstanceId, sequenceId, sequenceId + result.size(), false, result);
+        }
+
+        /**
+         * Drops pages up to the specified sequence id
+         */
+        private void acknowledgePages(long sequenceId)
+        {
+            checkState(!Thread.holdsLock(this), "Can not acknowledge pages while holding a lock on this");
+
+            synchronized (this) {
+                // if pages have already been acknowledged, just ignore this
+                long oldCurrentSequenceId = currentSequenceId.get();
+                if (finished.get() || sequenceId < oldCurrentSequenceId) {
+                    return;
+                }
+
+                int pagesToRemove = Ints.checkedCast(sequenceId - oldCurrentSequenceId);
+                checkState(pagesToRemove <= pages.size(), "Invalid sequence id");
+                long bytesRemoved = 0;
+                for (int i = 0; i < pagesToRemove; i++) {
+                    Page page = pages.removeFirst();
+                    bytesRemoved += page.getSizeInBytes();
+                }
+
+                verify(currentSequenceId.compareAndSet(oldCurrentSequenceId, oldCurrentSequenceId + pagesToRemove));
+
+                bufferedBytes.addAndGet(-bytesRemoved);
+                verify(bufferedBytes.get() >= 0);
+                memoryManager.updateMemoryUsage(-bytesRemoved);
+
+                // if there are no pages, and there will be no more pages, then we are finished
+                // never finish a failed buffer
+                if (pages.isEmpty() && !state.get().canAddPages() && state.get() != FAILED) {
+                    finished.set(true);
```

Comment:
The finished flag is generally an optimization, but is required to make the `getBufferInfo` method lock free.